libnmstate/nm/sriov.py

```diff
@@ -43,11 +90,67 @@ def create_setting(iface_state, base_con_profile):
         if not sriov_setting:
             sriov_setting = nmclient.NM.SettingSriov.new()
 
+        vfs_config = sriov_config.get(Ethernet.SRIOV.VFS_SUBTREE, [])
+        vf_object_ids = {vf.get_index() for vf in sriov_setting.props.vfs}
+        vf_config_ids = {
+            vf_config[Ethernet.SRIOV.VFS.ID] for vf_config in vfs_config
+        }
+        _edit_sriov_vfs_in_setting(
+            vfs_config, sriov_setting, vf_config_ids & vf_object_ids
+        )
+        _create_sriov_new_vfs_in_setting(
+            vfs_config, sriov_setting, vf_config_ids - vf_object_ids
+        )
+        _remove_sriov_vfs_in_setting(
+            vfs_config, sriov_setting, vf_object_ids - vf_config_ids
+        )
+
         sriov_setting.props.total_vfs = sriov_config[Ethernet.SRIOV.TOTAL_VFS]
 
     return sriov_setting
 
 
+def _create_sriov_new_vfs_in_setting(vfs_config, sriov_setting, vf_ids_to_add):
+    vfs_config_to_add = (
+        vf_config
+        for vf_config in vfs_config
+        if vf_config[Ethernet.SRIOV.VFS.ID] in vf_ids_to_add
+    )
+    for vf_config in vfs_config_to_add:
+        vf_object = nmclient.NM.SriovVF.new(vf_config[Ethernet.SRIOV.VFS.ID])
+        for key, val in vf_config.items():
+            _set_nm_attribute(vf_object, key, val)
+
+        sriov_setting.add_vf(vf_object)
+
+
+def _edit_sriov_vfs_in_setting(vfs_config, sriov_setting, vf_ids_to_edit):
+    vfs_config_to_edit = (
+        vf_config
+        for vf_config in vfs_config
+        if vf_config[Ethernet.SRIOV.VFS.ID] in vf_ids_to_edit
+    )
+    for vf_config in vfs_config_to_edit:
+        vf_id = vf_config[Ethernet.SRIOV.VFS.ID]
+        vf_object = nmclient.NM.SriovVF.new(vf_config[Ethernet.SRIOV.VFS.ID])
+        for key, val in vf_config.items():
+            _set_nm_attribute(vf_object, key, val)
+
+        sriov_setting.remove_vf_by_index(vf_id)
+        sriov_setting.add_vf(vf_object)
+
+
+def _set_nm_attribute(vf_object, key, value):
+    if key != Ethernet.SRIOV.VFS.ID:
```

Comment:
Can you pop this out once per `vf_config` at the caller? It looks wasteful to ask this for each attribute.

Reply:
Done! Thanks!